### DISCUSSION TITLE
Require `exactOptionalPropertyTypes`

### DISCUSSION THREAD

**Discussion by tjjfvi** [typed-ember/semver#5] [General]
Just as requiring `noUncheckedIndexedAccess` is useful to ensure compatibility with downstream code, it may make sense to also require `exactOptionalPropertyTypes`.

**Comment by chriskrycho** [comment]
I go back and forth on `exactOptionalPropertyTypes` personally, but your point about downstream code is an important one. @dfreeman @jamescdavis I would love your thoughts here as well!

**Comment by jamescdavis** [comment]
I've been using `exactOptionalPropertyTypes` on projects and I'm liking its explicitness. Especially for cases where objects are spread (or `Object.assign`ed), it's really nice to be able to allow a property to be overridden, but not cleared, e.g. https://tsplay.dev/wXKrDW

But regardless of my personal opinion, I agree it should be required to ensure downstream compatibility.

**Comment by chriskrycho** [comment]
@dfreeman and I were discussing this this morning and he pointed out that it's actually a bit *weirder* than "just require it" allows for: in fact, what is "safe" for a consumer *depends on what the library has chosen*, because of the way `exactOptionalPropertyTypes` interacts with variance.[^v]

Consider the following two examples, with `exactOptionalPropertyTypes` and function arguments and return types respectively ([playground][p]):

```ts
function exampleIn(arg: { foo?: string }) {
  if ('foo' in arg) {
    arg.foo.length;
  }
}
```

```ts
declare function exampleOut(): { foo?: string });
```

In the `exampleIn` case, if the library has `exactOptionalPropertyTypes` on, then changing from `{ foo?: string }` to `{ foo?: string | undefined }` changes what is legal for a caller to *pass*, because the `arg.foo.length` check no longer passes: the presence or absence of `foo` within the object no longer specifies that it is a `string`. But callers do not know this unless they *also* have `exactOptionalPropertyTypes` on: with it off, they can explicitly pass `{ foo: undefined }` in either case.

The exact inverse is true for the `exampleOut` case: explicitly returning `{ foo: undefined }` will break the same `in`-based narrowing done by a caller who is working with the return type.

(Notably, the check is an important one: the type system is catching real runtime issues!)

The only way that this can be safe is not by requiring `exactOptionalPropertyTypes` in libraries but rather by requiring libraries to *specify* which setting they use and informing consumers that they will only be safe if they *share* the choice made by the libraries they consume. The obvious problem here is: if you consume two libraries with different settings for this, you're out of luck. So… maybe we need to (a) specify that libraries should *always* use `exactOptionalPropertyTypes` *and* (b) specify that consumers must do so as well *or* risk breakage. That feels not great, but… 🤷🏼‍♂️ 

[p]: https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/HYQwtgpgzgDiDGEAEBBJBvAUEpEAeMA9gE4AuSAZgK7DykCWhwue4MANhAJLAAUIxAOYAuDJUKEA-KKili9YIKQBfAJQZsOJPQpJeAcgoT925gMHqsWreYB0Rwrc6LSACwDcmnMs0-N+IjIkABMIeHYBZGpaBiYWNk4AeSpSXlVRdHEpGTkFJQAfJBpQigUIYJVPP1BIWARkACENHACScmi6RmZ8BO4+cwys6SRZeUUkQuKIUuBylUsvbV0DBxMFJHMF6xw7BycIFw9Fn29MfwI2kLCI4iiaTriesA4IZNT0sQdh0bzKzD8ULYni8eLx0GpPIDgZxQZkHKIpjM5hDMFDWM8YXw4RJRPpXBB2OxCCYUZgGkD0SCsSjydC+mCsgjgCUyhUaRTerDGUg8QSiSTVJ5MJxyCAkABeVAcjGvFJpTw6PSGYymDZbDb2CT7Q5VM4ipAAIwlSFplKScsFmEVKxV6wN6oNmsczkEbl1QA

[^v]: Pretty sure "the way <feature> interacts with variance" is going to become my most-hated phrase of all time, because it is *«checks notes»* basically the root of *all* the most complicated pieces of this.

**Comment by tjjfvi** [comment]
Huh, good point. What about: require `exactOptionalPropertyTypes` in libraries, and require `| undefined` in user-constructible optional properties. So `exampleIn` would be invalid, as it would need to be:
```ts
function exampleIn(arg: { foo?: string | undefined }) {
  if ('foo' in arg) {
    arg.foo.length;
  }
}
```

**Reply by chriskrycho** [reply]
Huh, that seems like a pretty good solution, yeah. In practice, I don’t find the distinction *useful* as a library author, just… necessary, because JavaScript. 😂 Requiring a library to explicitly say "you can either explicitly or implicitly pass me nothing and I'll handle it" seems 👍🏼 .

**Comment by chriskrycho** [comment]
Long-delayed follow-up: this has been integrated since either beta.3 or the (just-published) beta.4. 👍🏼
